lib/puppet_x/libkv/file_plugin.rb

```diff
@@ -0,0 +1,315 @@
+    if File.directory?(key_file)
+      success = false
+      err_msg = "libkv plugin #{@name}: Key specifies a folder"
+    else
+      begin
+        File.unlink(key_file)
+        success = true
+      rescue Errno::ENOENT
+        # if the key doesn't exist, doesn't need to be deleted...going
+        # to consider this success
+        success = true
+      rescue Exception => e
+        success = false
+        err_msg = "Delete failed: #{e.message}"
+      end
+    end
+
+    { :result => success, :err_msg => err_msg }
+  end
+
+  # Deletes a whole folder from the configured backend.
+  #
+  # @param keydir String key folder path
+  #
+  # @return results Hash
+  #   * :result - Boolean indicating whether operation succeeded
+  #   * :err_msg - String. Explanatory text upon failure; nil otherwise.
+  #
+  def deletetree(keydir)
+    success = nil
+    err_msg = nil
+    dir = File.join(@root_path, keydir)
+    # FIXME:  Is there an atomic way of doing this?
+    if Dir.exist?(dir)
+      begin
+        FileUtils.rm_r(dir)
+        success = true
+      rescue Exception => e
+        if Dir.exist?(dir)
+          success = false
+          err_msg = "Folder delete failed: #{e.message}"
+        else
+          # in case another process/thread successfully deleted the directory
+          success = true
+        end
+      end
+    else
+      # if the directory doesn't exist, doesn't need to be deleted...going
+      # to consider this success
+      success = true
+    end
+
+    { :result => success, :err_msg => err_msg }
+  end
+
+  # Returns whether the `key` exists in the configured backend.
+  #
+  # @param key String key
+  #
+  # @return results Hash
+  #   * :result - Boolean indicating whether key exists; nil if could not
+  #     be determined
+  #   * :err_msg - String. Explanatory text when status could not be
+  #     determined; nil otherwise.
+  #
+  def exists(key)
+    key_file = File.join(@root_path, key)
+    # this simple plugin doesn't have any error cases that would be reported
+    # in :err_msg
+    { :result => File.exist?(key_file), :err_msg => nil }
+  end
+
+  # Retrieves the value stored at `key` from the configured backend.
+  #
+  # @param key String key
+  #
+  # @return results Hash
+  #   * :result - String. Retrieved value for the key; nil if could not
+  #     be retrieved
+  #   * :err_msg - String. Explanatory text upon failure; nil otherwise.
+  #
+  def get(key)
+    value = nil
+    err_msg = nil
+    key_file = File.join(@root_path, key)
+    if File.directory?(key_file)
+      err_msg = "libkv plugin #{@name}: Key specifies a folder"
+    else
+      begin
+        Timeout::timeout(@lock_timeout_seconds) do
+          # To ensure all threads are not sharing the same file descriptor
+          # do **NOT** use a File.open block!
+          file = File.open(key_file, 'r')
+          file.flock(File::LOCK_EX)
+          value = file.read
+          file.close # lock released with close
```

Comment:
Do you want to release the lock and close on a `Timeout` exception? I don't think that this will happen normally.

Reply:
:+1: Good catch

Reply:
Logic reworked to address lock release issue.